### ISSUE TITLE
h2childchan: Performance

### ISSUE THREAD

**Issue by buchgr** [netty/netty#5772]
This is an issue to track the performance of the [HTTP/2 child channel ](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Codec.java)API. We need to get it close to the [connection-based](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java) API.

I ran some benchmarks aiming for max throughput. I modified the hello world examples from 4.1-dev. Removed HTTP Upgrade and disabled logging. I allocated two cloud virtual machines with 8 cores. I ran the HTTP/2 server on one host. The [h2load](https://nghttp2.org/documentation/h2load-howto.html) tool on the other host.

The below is the best out of 3 runs. Results were pretty stable between runs. The connection-based API has a roughly 30% higher throughput (640k req/s vs 495k req/s).

**Connection-based API**

```
buchgr@buchgr-vm-1:~$ ./h2load -n1000000 -c8 -m100 http://10.0.0.6:8080
starting benchmark...
spawning thread #0: 8 total client(s). 1000000 total requests
Application protocol: h2c

finished in 1.56s, 639898.44 req/s, 29.29MB/s
requests: 1000000 total, 1000000 started, 1000000 done, 1000000 succeeded, 0 failed, 0 errored, 0 timeout
status codes: 1000000 2xx, 0 3xx, 0 4xx, 0 5xx
traffic: 45.78MB (48000144) total, 976.56KB (1000000) headers (space savings 90.00%), 22.89MB (24000000) data
                     min         max         mean         sd        +/- sd
time for request:      598us      7.76ms      1.18ms       284us    89.91%
time for connect:      834us      1.67ms       987us       260us    87.50%
time to 1st byte:     3.27ms      4.95ms      4.36ms       504us    75.00%
req/s           :   80024.48    80656.84    80330.29      201.66    62.50%
```

**Child Channel API**

```
buchgr@buchgr-vm-1:~$ ./h2load -n1000000 -c8 -m100 http://10.0.0.6:8080
starting benchmark...
spawning thread #0: 8 total client(s). 1000000 total requests
Application protocol: h2c

finished in 2.02s, 494949.29 req/s, 17.46MB/s
requests: 1000000 total, 1000000 started, 1000000 done, 1000000 succeeded, 0 failed, 0 errored, 0 timeout
status codes: 1000000 2xx, 0 3xx, 0 4xx, 0 5xx
traffic: 35.29MB (37000144) total, 976.56KB (1000000) headers (space savings 90.00%), 12.40MB (13000000) data
                     min         max         mean         sd        +/- sd
time for request:      319us     44.21ms      1.46ms      1.18ms    97.68%
time for connect:      832us      1.03ms       934us        75us    50.00%
time to 1st byte:     4.13ms      8.29ms      5.22ms      1.19ms    87.50%
req/s           :   61890.26    68507.89    66076.86     2179.42    62.50%
```

cc: @ejona86 @nmittler @louiscryan @carl-mastrangelo @Scottmitch @normanmaurer @trustin @mosesn @vietj

**Comment by mosesn** [comment]
Hmm, interesting.  Have you profiled it?  Any ideas why this might be the case?

**Comment by carl-mastrangelo** [comment]
Happen to have the CPU utilization for the SUT?  My own profiling has shown small amounts of lock contention in Netty's DefaultResourceLeak.<init>  which might be created pretty fast.

**Comment by normanmaurer** [comment]
@carl-mastrangelo yeah there is some locking overhead in `DefaultResourceLeak` which could most likely be removed by CAS (there is even a TODO for it I think :)). That said you could just disable leak-detection to see if this is really the cut-pit here. I suspect it is not.

**Comment by buchgr** [comment]
I didn't do any profiling. Will look tomorrow. Generally, we expect the child channel based API to have a lower throughput, as it builds on top of the connection based API. So we can never be faster - we can just aim to not be much slower :).

Some of the overhead from the top of my head: The child channel API uses two additional channel handlers on the parent channel. For every frame it has to allocate an additional object. For every stream it allocates an additional channel, pipeline, config, etc. For every write it has to allocate an additional promise, as promises on the child channel can't be used on the parent channel. There is some overhead in the multiplexing logic. Every child channel allocates a new object to the tell the parent channel that it consumed the bytes for flow control. When a stream goes active / inactive a message between the frame codec and multiplex codec is exchanged.

However, I have neither done any profiling nor looked at the GC logs or aimed at efficiency when working on the child channel API. I don't think @ejona86 did either. So there is hope! I think the goal for now should be to get it as close as possible to the connection based API (say within 10% of the throughput). If successful, we can look into merging the connection based API and the child channel API. We should be able to remove a bit of overhead from the connection based API. For example, the `Http2Stream` objects would become the child channels. We would no longer need the properties on the `Http2Stream`, but have them as fields on the child channel directly. Probably a bunch of other stuff too.

Thoughts?

**Comment by mosesn** [comment]
Makes a lot of sense to me.  Thanks @buchgr.  I also wonder if we might be running into some unnecessary lock concurrency?  Hopefully profiling will show up some of that.

**Comment by buchgr** [comment]
Right, we wouldn't expect any additional lock contention on child channels. I wouldn't worry too much about `DefaultResourceLeak`, as @normanmaurer noted it can simply be disabled. My rule of thumb is that CASing is nice if you don't have contention, but more expensive than a OS lock if you do. Also replacing a lock with a CAS spin-loop will just not show as lock contention in your profiler ... so it's a good way to get rid of lock contention 😄.

**Comment by buchgr** [comment]
I had a look at the performance. I first looked at garbage collection. 

I ran 10 million requests `./h2load -n10000000 -c8 -m100 http://10.0.0.6:8080` and summed up GC STW + TTS pauses.

I ran the benchmark 3 times. 

**Connection-based**
```
benchmark duration: 17.55s, STW Pauses: 0.21167s (1.2%)
benchmark duration: 16.44s, STW Pauses: 0.162271s (1%)
benchmark duration: 16.21s, STW Pauses: 0.180943s (1.1%)
```

**Child Channel**
```
benchmark duration: 21.16s, STW Pauses: 0.457769s (2.1%)
benchmark duration: 20.91s, STW Pauses: 0.504049s (2.4%)
benchmark duration: 22.65s, STW Pauses: 0.554505s (2.4%)
```

It seems that with child channels, the total time spent in pauses roughly doubles. Individual GC times don't seem higher, there are just more GC events (only young, no old, old gen is only 4% occupied). That's bad. However, even then the application spends only ~2.5% of it's time in GC. Which doesn't explain a ~23% slowdown.

Looking at the mean request-response latency it's 4.36ms (connection) vs. 5.22ms (child channel). Assuming the same network latency, it seems plausible that child channels do `850 micros` more work per request.

I ll look at the CPU profile next.

**Comment by buchgr** [comment]
CPU profiling with Honest Profiler and Java Flight Recorder doesn't paint a useful picture. We don't have individual methods consuming large amounts of CPU. Results between profiling sessions weren't stable either. According to those profilers we have some 20 methods consuming 2 - 3% of CPU, the rest is less than 1%. So nothing that pops in the eye, or that could be heavily optimized. 

So I turned to flame graphs. In order to remove noise and overhead, I sampled only a single eventloop thread with very high resolution `perf record -F 997 -t thread-id -g -- sleep 60` for 1 minute. After all, in this benchmark all eventloop threads should do about the same work.

Here is the flamegraph: [multiplex.svg.zip](https://github.com/netty/netty/files/451963/multiplex.svg.zip)

The flamegraph summary is: We spend 26% of CPU time in child channels code. 10% on creating, initializing and registering new channels. 16% on reading (7%), writing (7%), and closing channels (2%).

From a first look, it seems that we don't do anything stupid that consumes tons of time and that could be easily optimized. It seems we just do more work.

**Comment by buchgr** [comment]
Profiling has shown that virtually all of the overhead is due to child channels. That is, the `Http2MultiplexCodec`. In order to be sure, I implemented a HTTP/2 server using only the `Http2FrameCodec` https://github.com/buchgr/netty/commit/6d4a6e3bc693805f91f1cebe7fdbdf007b22ab9d. The frame codec sits on top of the connection based API and essentially replaces callbacks by messages.

The below is the best out of 3 runs. Throughput and latency seem identical. Also the GC times are practically identical (about 1% or 0.22s per benchmark).

**Message Passing / Http2FrameCodec**
```
buchgr@buchgr-vm-1:~$ ./h2load -n10000000 -c8 -m100 http://10.0.0.6:8080
starting benchmark...
spawning thread #0: 8 total client(s). 10000000 total requests
Application protocol: h2c

finished in 15.28s, 654633.88 req/s, 29.97MB/s
requests: 10000000 total, 10000000 started, 10000000 done, 10000000 succeeded, 0 failed, 0 errored, 0 timeout
status codes: 10000000 2xx, 0 3xx, 0 4xx, 0 5xx
traffic: 457.76MB (480000144) total, 9.54MB (10000000) headers (space savings 90.00%), 228.88MB (240000000) data
                     min         max         mean         sd        +/- sd
time for request:      245us     24.52ms      1.15ms       545us    93.12%
time for connect:      964us      1.14ms      1.04ms        66us    50.00%
time to 1st byte:     6.50ms     20.97ms     12.37ms      6.66ms    62.50%
req/s           :   81833.68    83049.28    82367.57      398.25    50.00%

```

**Connection-based**
```
buchgr@buchgr-vm-1:~$ ./h2load -n10000000 -c8 -m100 http://10.0.0.6:8080
starting benchmark...
spawning thread #0: 8 total client(s). 10000000 total requests
Application protocol: h2c

finished in 15.35s, 651362.17 req/s, 29.82MB/s
requests: 10000000 total, 10000000 started, 10000000 done, 10000000 succeeded, 0 failed, 0 errored, 0 timeout
status codes: 10000000 2xx, 0 3xx, 0 4xx, 0 5xx
traffic: 457.76MB (480000144) total, 9.54MB (10000000) headers (space savings 90.00%), 228.88MB (240000000) data
                     min         max         mean         sd        +/- sd
time for request:      491us     12.63ms      1.16ms       390us    94.12%
time for connect:      711us       873us       804us        63us    37.50%
time to 1st byte:     5.33ms     10.36ms      7.80ms      1.64ms    50.00%
req/s           :   81424.86    82356.68    81794.46      308.37    75.00%
```

FYI @nmittler @ejona86

**Comment by normanmaurer** [comment]
@buchgr thats interesting... so at least we know now that its not the "message conversation" that is the problem. Any idea yet how to reduce the overhead of child-channels ?

**Comment by normanmaurer** [comment]
@buchgr after looking in the child channel impl I think a lot of overhead comes from the fact that it uses `AbstractChannel`. I may have a few hour on my way to SFO to hack on it... Maybe I can remove some of the overhead...

**Comment by buchgr** [comment]
@normanmaurer 

No. We probably can't reduce it _significantly_. Channel creation, Channel events, Pipeline traversal, etc. just take their toll. That being said, the child channels are still _very_ fast. We have to keep in mind that in this benchmark, the application logic is about 2% of total CPU time. In a real application, it will be closer to 60 - 80%, making the slowdown unnoticable for all but the most performance sensitive applications.

After discussions with @ejona86 and @nmittler, it seems that a viable way forward is to focus on integrating the `Http2FrameCodec` and the connection based API, so that we can remove/deprecate the connection based API and mark the `Http2FrameCodec` API stable. The `Http2MultiplexCodec` / child channels are completely independent of the `Http2FrameCodec` and can be viewed as an add-on that makes development really nice and easy, although with a small performance penalty. Most Netty users, will want to build their applications on top of the `Http2MultiplexCodec`. Some really performance sensitive, like gRPC will choose the `Http2FrameCodec` and do their own multiplexing. If you look at how I implemented the HTTP/2 server on top of the `Http2FrameCodec` [1], it looks and feels like Netty in my opinion.

Thoughts?

[1] https://github.com/buchgr/netty/commit/6d4a6e3bc693805f91f1cebe7fdbdf007b22ab9d#diff-45724a8e3add16ceeda9ae0832fafd00R37

**Comment by buchgr** [comment]
@normanmaurer 

> @buchgr after looking in the child channel impl I think a lot of overhead comes from the fact that it uses AbstractChannel. I may have a few hour on my way to SFO to hack on it... Maybe I can remove some of the overhead...

Any specifics that you would change? If you look at the flamegraph, the `AbstractChannel` doesn't show up much. Only maybe channel registration / initialization.

**Comment by normanmaurer** [comment]
@buchgr yeah I agree even without the child channel stuff it looks very nice and netty typically. Its very close to the SPDY codec as well :) So yep I agree even without it is a huge win! Let me see if I can come up with some improvement for the child channels thought...

Yeah I I think we can reduce some overhead by doing some changes for the channel registration / init and also by remove the extra outbound buffer.

**Comment by buchgr** [comment]
I made a big mistake when benchmarking.  @normanmaurer noticed that something was off when running the benchmarks on his laptop, as he couldn't reproduce my results. I then found that the load client used only a single thread and didn't saturate the server. So the above results are at ~60% CPU load, not maximum throughput. I am very sorry.

I re-ran the benchmarks (with `./h2load -n20000000 -c128 -m10000 -t8 http://10.0.0.6:8080`), double checking that the server CPU is saturated. The `Http2FrameCodec` is `25%` slower. The `Http2FrameCodec + Http2MultiplexCodec` is `64%` slower.

**Connection-based API**
```
1300518.82 req/s, 9.08ms mean latency,  0.9% STW pauses
```

**Multiplex/Child Channel API**
```
474339.60 req/s, 25.22ms mean latency, 0.9% STW pauses
```

**Message API**
```
964726.33 req/s, 11.71ms mean latency, 0.8% STW pauses
```
